### PR TITLE
Implement job function for Commit Pool - Closes #6850

### LINKED ISSUE #6850: Implement job function for CommitPool
### Description

Implement `job()` function of CommitPool as per specified in [LIP:SingleCommit:Gossip](https://research.lisk.com/t/introduce-certificate-generation-mechanism/296#single-commit-p2p-gossip-37),

- Periodic Job Triggered every `BLOCK_TIME/2`, Cleans up data structures and gossips the commit messages based on condition
- Remove stale or invalid commit messages
- Choose commit messages according to [LIP:SingleCommit:Gossip](https://research.lisk.com/t/introduce-certificate-generation-mechanism/296#single-commit-p2p-gossip-37)
- Gossip commit messages by calling send() of network to 16 randomly chosen connected peers
- Move any gossiped commit messages from `_nonGossipedCommits` to `_gossipedCommits`

### Acceptance Criteria

- Add unit test for `job()`

### PR DESCRIPTION
**Pull request by ishantiw**
### What was the problem?

This PR resolves #6850

### How was it solved?

🌱 Implement job function for commitPool 365226b47455d56378ec6d253787054e3d3dbac4
✅ Add tests for Job function 52bc919e24a66dce52eddf91c803e690d80d0e70

### How was it tested?

Run `npm run test:unit commit_pool` under `framework`

### CODE REVIEW

File: framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts [hunk 1]
```diff
@@ -82,22 +86,217 @@ describe('CommitPool', () => {
...
+			// Act
+			await commitPool['_job'](context);
+			// Assert
+			// nonGossiped commits are moved to gossiped commits
+			expect([...commitPool['_nonGossipedCommits'].values()].flat()).toHaveLength(0);
+			expect([...commitPool['_gossipedCommits'].values()].flat()).toHaveLength(10);
+			expect(commitPool['_nonGossipedCommits'].get(1070)).toBeUndefined();
+		});
+
+		it('should clean all the commits from gossipedCommit that does not have bftParams change', async () => {
```
Comment (Incede) [inline framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts:231]: I think this test is redundant (same arrange, act, assert) and can be merged with the previous (L202)

File: framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts [hunk 2]
```diff
@@ -82,22 +86,217 @@ describe('CommitPool', () => {
...
+			await commitPool['_job'](context);
+			// Assert
+			// nonGossiped commits are moved to gossiped commits
+			expect([...commitPool['_nonGossipedCommits'].values()].flat()).toHaveLength(0);
+			expect([...commitPool['_gossipedCommits'].values()].flat()).toHaveLength(11);
+			expect(commitPool['_nonGossipedCommits'].get(1070)).toBeUndefined();
+			const commits = commitPool['_gossipedCommits'].get(1070);
+			expect(commits).toBeDefined();
+			expect(commits).toBeArray();
+			const generatorCommit = commits?.find(c => c.validatorAddress.equals(generatorAddress));
```
Comment (Incede) [inline framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts:286]: This commit by generator is under `_gossipedCommits` but as per the description it says `select non gossiped commits`.
Maybe improve the description.

File: framework/src/node/consensus/certificate_generation/commit_pool.ts [hunk 3]
```diff
@@ -372,10 +384,160 @@ export class CommitPool {
...
+
+		// 2. Select commits to gossip
+		const validators = await this._bftAPI.getCurrentValidators(apiContext);
+		const numActiveValidators = validators.length;
+		// Get a list of commits sorted by ascending order of height
+		const allCommits = this._getAllCommits();
+
+		const selectedCommits = [];
+		for (const commit of allCommits) {
+			if (selectedCommits.length >= numActiveValidators) {
```
Comment (Incede) [inline framework/src/node/consensus/certificate_generation/commit_pool.ts:462]: I think this condition is not being tested by unit test.

File: framework/src/node/consensus/certificate_generation/commit_pool.ts [hunk 4]
```diff
@@ -372,10 +384,160 @@ export class CommitPool {
...
+				break;
+			}
+
+			// 2.1 Choosing the commit with higher height first
+			if (commit.height < maxHeightPrecommitted - COMMIT_RANGE_STORED) {
```
Comment (Incede) [inline framework/src/node/consensus/certificate_generation/commit_pool.ts:467]: maybe this too ? but not sure

File: framework/src/node/consensus/certificate_generation/commit_pool.ts [hunk 5]
```diff
@@ -372,10 +384,160 @@ export class CommitPool {
...
+			}
+		}
+		// Non gossiped commits with descending order of height
+		const sortedNonGossipedCommits = [...this._nonGossipedCommits.values()]
+			.flat()
+			.sort((a, b) => b.height - a.height);
+
+		// 2.2 Select newly created commits by generator
+		for (const [index, commit] of sortedNonGossipedCommits.entries()) {
+			if (selectedCommits.length >= numActiveValidators) {
```
Comment (Incede) [inline framework/src/node/consensus/certificate_generation/commit_pool.ts:478]: same

**Review by Incede — CHANGES_REQUESTED** [review]

File: framework/src/node/consensus/certificate_generation/commit_pool.ts [hunk 6]
```diff
@@ -372,10 +384,160 @@ export class CommitPool {
...
+		}
+
+		const encodedCommitArray = selectedCommits.map(commit =>
+			codec.encode(singleCommitSchema, commit),
+		);
+		// 3. Gossip an array of up to 2*numActiveValidators commit messages to 16 randomly chosen connected peers with at least 8 of them being outgoing peers (same parameters as block propagation)
+		this._network.send({
+			event: EVENT_COMMIT_MESSAGES,
+			data: codec.encode(singleCommitsNetworkPacketSchema, { commits: encodedCommitArray }),
+		});
```
Comment (shuse2) [inline framework/src/node/consensus/certificate_generation/commit_pool.ts:527]: sent commits should be moved from non gossiped to gossiped here?

File: framework/src/node/consensus/certificate_generation/constants.ts [hunk 7]
```diff
@@ -15,3 +15,4 @@
 export const MESSAGE_TAG_CERTIFICATE = 'LSK_CE_';
 export const COMMIT_RANGE_STORED = 100;
 export const EMPTY_BUFFER = Buffer.alloc(0);
+export const EVENT_COMMIT_MESSAGES = 'EVENT_COMMIT_MESSAGES';
```
Comment (shuse2) [inline framework/src/node/consensus/certificate_generation/constants.ts:18]: probably this should be consistent with 
```
export const NETWORK_EVENT_POST_BLOCK = 'postBlock';
```

**Review by shuse2 — CHANGES_REQUESTED** [review]
Checked the implementation part. once resolved, ill review the test

File: framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts [hunk 8]
```diff
@@ -82,22 +86,217 @@ describe('CommitPool', () => {
...
+			commitPool['_bftAPI'].existBFTParameters = jest.fn().mockResolvedValue(true);
+			const context = createNewAPIContext(new InMemoryKVStore());
+			// Act
+			await commitPool['_job'](context);
+			// Assert
+			// nonGossiped commits are moved to gossiped commits and stale commit is deleted
+			expect([...commitPool['_nonGossipedCommits'].values()].flat()).toHaveLength(0);
+		});
+
+		it('should clean all the commits from gossipedCommit list with height below removal height', async () => {
```
[the remark this answers is not in the document]
Comment (ishantiw) [reply framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts:189]: commits are also deleted but yes they are also added so the desc is misleading

File: framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts (hunk 1 above)
Comment (ishantiw) [reply framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts:231]: they are different, the previous is adding failing BFT param commit to nonGossiped and this one in gossiped and then later we are checking for commits presence in the list it was added. But code seems very redundant so I have merged both the tests in one

File: framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts (hunk 2 above)
Comment (ishantiw) [reply framework/test/unit/node/consensus/certificate_generation/commit_pool.spec.ts:286]: good catch 😉

File: framework/src/node/consensus/certificate_generation/commit_pool.ts (hunk 4 above)
Comment (ishantiw) [reply framework/src/node/consensus/certificate_generation/commit_pool.ts:467]: all these conditions are checked in the same unit test created for above one

File: framework/src/node/consensus/certificate_generation/commit_pool.ts [hunk 9]
```diff
@@ -372,10 +384,160 @@ export class CommitPool {
...
+		const sortedNonGossipedCommits = [...this._nonGossipedCommits.values()]
+			.flat()
+			.sort((a, b) => b.height - a.height);
+
+		// 2.2 Select newly created commits by generator
+		for (const [index, commit] of sortedNonGossipedCommits.entries()) {
+			if (selectedCommits.length >= numActiveValidators) {
+				break;
+			}
+			if (commit.validatorAddress.equals(this._generatorAddress)) {
```
[the remark this answers is not in the document]
Comment (ishantiw) [reply framework/src/node/consensus/certificate_generation/commit_pool.ts:419]: yes that will be fixed in the issue https://github.com/LiskHQ/lisk-sdk/issues/6969

File: framework/src/node/consensus/certificate_generation/constants.ts [hunk 10]
```diff
@@ -15,4 +15,4 @@
 export const MESSAGE_TAG_CERTIFICATE = 'LSK_CE_';
 export const COMMIT_RANGE_STORED = 100;
 export const EMPTY_BUFFER = Buffer.alloc(0);
-export const EVENT_COMMIT_MESSAGES = 'EVENT_COMMIT_MESSAGES';
+export const NETWORK_EVENT_COMMIT_MESSAGES = 'commitMessages';
```
Comment (shuse2) [inline framework/src/node/consensus/certificate_generation/constants.ts:18]: `Message` seems too generic. maybe `postSingleCommitMessages`?

**Review by shuse2 — CHANGES_REQUESTED** [review]

File: framework/src/node/consensus/certificate_generation/commit_list.ts [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+		return this._commitMap.keys();
+	}
+
+	public add(commit: SingleCommit) {
+		const currentCommits = this._commitMap.get(commit.height) ?? [];
+
+		this._commitMap.set(commit.height, [...currentCommits, commit]);
+	}
+
+	public commitExist(commit: SingleCommit) {
```
[the remark this answers is not in the document]
Comment (shuse2) [reply framework/src/node/consensus/certificate_generation/commit_list.ts]: This goes for other methods, such as `getCommitsByHeight`

**Review by shuse2 — APPROVED** [review]
Minor naming things but LGTM 🎉

**Review by Incede — APPROVED** [review]

**Pull request merged**